# Segmenting and Clustering Neighbourhoods in Toronto
### First of all, let's import the pandas library and read the html of the Wikipedia url.

In [102]:
import pandas as pd

pd.read_html('https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=1011037969')

dfs = pd.read_html('https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=1011037969', header=0)

### By using [0] as our header, we can now determine a dataframe called 'df'.

In [103]:
df = dfs[0]
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

As we can see above, there's a lot of 'Not assigned' values. We need to fix that!

### To fix these values, we will simply create a filtered dataframe which will not have any 'Not assigned' values.

In [104]:
df_filtered = df[df['Borough'] != 'Not assigned'] 

In [105]:
df_filtered.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

It is clear that the neighbourhoods are already grouped by the Postal Code, so there isn't any 'Not assigned' values in the 'Neighbourhood' column.

### Now we can finally check the dataframe shape by using the .shape function!

In [106]:
df_filtered.shape

(103, 3)

So the new dataframe has 103 rows and 3 columns!

### Let's import the Geospatial Coordinates data.
We also need to check if it matches the same number of rows and columns from our filtered dataframe.

In [107]:
import os, types
import pandas as pd
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# @hidden_cell
import os, types
import pandas as pd
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# @hidden_cell

# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_1 = pd.read_csv(body)
df_data_1.shape

(103, 3)

As shown by the .shape function, the Geospatial_Coordinates.csv has 103 rows and 3 columns too! It matches the filtered dataframe .shape function.

### We can proceed to merge the two dataframes.
Let's use the .merge() function!

In [108]:
neighbourhoods = df_filtered.merge(df_data_1,left_on='Postal Code',right_on='Postal Code')

In [109]:
neighbourhoods.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

There it is! The merged table is ready for use!

# Let's make sure to import and install the necessary libraries!

In [110]:
import numpy as np # library to handle data in a vectorized manner

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!pip install geopy

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors


!pip install folium
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## We'll begin by defining the address as Toronto and getting its coordinates

In [111]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="on_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))


The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


## We are now able to create a map from the information above

In [112]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, lng, borough, neighbourhood in zip(neighbourhoods['Latitude'], neighbourhoods['Longitude'], neighbourhoods['Borough'], neighbourhoods['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  

map_toronto

## Let's choose a Borough to work with!
I've chosen Downtown Toronto, as it has good potential to be worked with.

In [113]:
toronto_data = neighbourhoods[neighbourhoods['Borough'] == "Downtown Toronto"].reset_index(drop=True)
toronto_data.head()


Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

Defining the Downtown Toronto address.

In [114]:
address = 'Downtown Toronto, ON'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))


The geograpical coordinate of Downtown Toronto are 43.6563221, -79.3809161.


## Creating a map of Downtown Toronto

In [115]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto


In the map we just created we can observe only the Neighbourhood within Downtown Toronto are shown!

## Defining our Foursquare API credentials
We'll be using the Foursquare API to analyze venues in Downtown Toronto.

In [ ]:
#@ hidden cell
CLIENT_ID = # your Foursquare ID
CLIENT_SECRET = # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)


### Let's check the first Neighbourhood in the toronto_data dataframe.

In [117]:
toronto_data.loc[0, 'Neighbourhood']

'Regent Park, Harbourfront'

In [118]:
neighbourhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighbourhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighbourhood_name = toronto_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))


Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


## Let's get the Top 100 venues in Regent Park within a radius of 500 meters.

In [ ]:
LIMIT = 100
radius = 500

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)
url


In [ ]:
results = requests.get(url).json()
results

As shown by the results, we got a lot of information!

Time to filter it!

### Function that extracts the category of the venue

In [121]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']


### Cleaning the json and making it a pandas dataframe

In [122]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))


45 venues were returned by Foursquare.


/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  app.launch_new_instance()


### Creating the getNearbyVenues function

In [123]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)

        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)
        

## Let's create a new dataframe using the function we just created!

In [124]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighbourhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )


Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


In [125]:
print(toronto_venues.shape)
toronto_venues.head()


(1221, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Regent Park, Harbourfront                43.65426               -79.360636   
1  Regent Park, Harbourfront                43.65426               -79.360636   
2  Regent Park, Harbourfront                43.65426               -79.360636   
3  Regent Park, Harbourfront                43.65426               -79.360636   
4  Regent Park, Harbourfront                43.65426               -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0           Tandem Coffee       43.653559       -79.361809   
1        Roselle Desserts       43.653447       -79.362017   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0          Coffee Shop  
1               Bakery  
2  Distribution Center  
3           Restaurant  
4                  Spa

We got a lot of results! Let's see how many venues each Neighbourhood returned!

In [126]:
toronto_venues.groupby('Neighbourhood').count()


Neighbourhood Latitude  \
Neighbourhood                                                                
Berczy Park                                                             60   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      63   
Christie                                                                16   
Church and Wellesley                                                    72   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               63   
Queen's Park, Ontario Provincial Government                             31   
Regent Park, Harbourfront                                               45   
Richmond, Adelaide, King                                                94   
Rosedale                                                                 4   
St. James Town                                                          80   
St. James Town, Cabbagetown                                             45   
Stn A PO Boxes                                                         100   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          32   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Berczy Park                                                              60   
CN Tower, King and Spadina, Railway Lands, Harb...                       16   
Central Bay Street                                                       63   
Christie                                                                 16   
Church and Wellesley                                                     72   
Commerce Court, Victoria Hotel                                          100   
First Canadian Place, Underground city                                  100   
Garden District, Ryerson                                                100   
Harbourfront East, Union Station, Toronto Islands                       100   
Kensington Market, Chinatown, Grange Park                                63   
Queen's Park, Ontario Provincial Government                              31   
Regent Park, Harbourfront                                                45   
Richmond, Adelaide, King                                                 94   
Rosedale                                                                  4   
St. James Town                                                           80   
St. James Town, Cabbagetown                                              45   
Stn A PO Boxes                                                          100   
Toronto Dominion Centre, Design Exchange                                100   
University of Toronto, Harbord                                           32   

                                                    Venue  Venue Latitude  \
Neighbourhood                                                               
Berczy Park                                            60              60   
CN Tower, King and Spadina, Railway Lands, Harb...     16              16   
Central Bay Street                                     63              63   
Christie                                               16              16   
Church and Wellesley                                   72              72   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100   

## Analyzing Each Neighborhood

In [127]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()


Neighbourhood  Adult Boutique  Airport  Airport Food Court  \
0  Regent Park, Harbourfront               0        0                   0   
1  Regent Park, Harbourfront               0        0                   0   
2  Regent Park, Harbourfront               0        0                   0   
3  Regent Park, Harbourfront               0        0                   0   
4  Regent Park, Harbourfront               0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  BBQ Joint  \
0                    0                 0                   0          0   
1                    0                 0                   0          0   
2                    0                 0                   0          0   
3                    0                 0                   0          0   
4                    0                 0                   0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0           0           0       0     0    0                 0   
1           0           0       1     0    0                 0   
2           0           0       0     0    0                 0   
3           0           0       0     0    0                 0   
4           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Butcher  Café  Candy Store  \
0             0              0        0     0            0   
1             0              0        0     0            0   
2             0              0        0     0            0   
3             0              0        0     0            0   
4             0              0        0     0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0          

In [128]:
toronto_onehot.shape


(1221, 209)

In [129]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

toronto_grouped.shape


(19, 209)

## Let's print each neighborhood along with the top 5 most common venues


In [ ]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')


## Let's make it into a dataframe

In [131]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]


In [132]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

        # create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()


Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                Bakery              Pharmacy   
1      Airport Terminal           Coffee Shop              Boutique   
2    Italian Restaurant        Sandwich Place                  Café   
3                  Café                  Park            Baby Store   
4   Japanese Restaurant      Sushi Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Farmers Market    Seafood Restaurant              Beer Bar   
1                 Plane   Rental Car Location         Boat or Ferry   
2       Thai Restaurant           Salad Place       Bubble Tea Shop   
3           Candy Store             Nightclub           Coffee Shop   
4           Yoga Studio                   Pub           Men's Store   

      8th Most Common Venue 9th Most Common Venue      10th Most Common Venue  
0                Restaurant           Cheese Shop              Breakfast Spot  
1           Harbor / Marina      Sculpture Garden             Airport Service  
2              Burger Joint   Japanese Restaurant  Modern European Restaurant  
3                Restaurant    Italian Restaurant          Athletics & Sports  
4  Mediterranean Restaurant                 Hotel                     Gay Bar

## Clustering using k-Means

In [133]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 


array([0, 2, 0, 4, 0, 0, 0, 0, 0, 0], dtype=int32)

In [134]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!


Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0                  Café   
4  43.644771 -79.373306               0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                  Park                   Pub   
1      Sushi Restaurant     College Cafeteria           Yoga Studio   
2           Coffee Shop        Cosmetics Shop       Bubble Tea Shop   
3           Coffee Shop            Restaurant             Gastropub   
4          Cocktail Bar                Bakery              Pharmacy   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0               Theater             Breakfast Spot                  Café   
1                 Diner      Portuguese Restaurant                  Park   
2                  Café  Middle Eastern Restaurant   Japanese Restaurant   
3          Cocktail Bar             Cosmetics Shop    Italian Restaurant   
4        Farmers Market         Seafood Restaurant              Beer Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Mexican Restaurant        Chocolate Shop           Dessert Shop  
1           Music Venue    Mexican Restaurant    Japanese Restaurant  
2  Fast Food Restaurant             Bookstore       Ramen Restaurant  
3                  Park    Seafood Restaurant    Moroccan Restaurant  
4            Restaurant           Cheese Shop         Breakfast Spot

# Let's finally visualize the cluster maps!

In [135]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]


markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters


# Time to examine the Clusters

In [136]:
#Cluster 1
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0        Clothing Store   
3   Downtown Toronto               0                  Café   
4   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0           Coffee Shop   
8   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
10  Downtown Toronto               0           Coffee Shop   
12  Downtown Toronto               0                  Café   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue  3rd Most Common Venue  \
0                 Bakery                   Park   
1       Sushi Restaurant      College Cafeteria   
2            Coffee Shop         Cosmetics Shop   
3            Coffee Shop             Restaurant   
4           Cocktail Bar                 Bakery   
5     Italian Restaurant         Sandwich Place   
7                   Café             Restaurant   
8               Aquarium                  Hotel   
9                  Hotel                   Café   
10            Restaurant                   Café   
12           Coffee Shop  Vietnamese Restaurant   
15    Seafood Restaurant           Cocktail Bar   
16                   Pub                 Bakery   
17                  Café                  Hotel   
18   Japanese Restaurant       Sushi Restaurant   

            4th Most Common Venue 5th Most Common Venue  \
0                             Pub               Theater   
1                     Yoga Studio                 Diner   
2                 Bubble Tea Shop                  Café   
3                       Gastropub          Cocktail Bar   
4                        Pharmacy        Farmers Market   
5                            Café       Thai Restaurant   
7                             Gym        Clothing Store   
8                            Café   Sporting Goods Shop   
9                      Restaurant           Salad Place   
10                          Hotel                   Gym   
12  Vegetarian / Vegan Restaurant    Mexican Restaurant   
15            Japanese Restaurant                Bakery   
16                     Restaurant                  Café   
17                     Restaurant   Japanese Restaurant   
18                     Restaurant           Yoga Studio   

        6th Most Common Venue 7th Most Common Venue     8th Most Common Venue  \
0              Breakfast Spot                  Café        Mexican Restaurant   
1       Portuguese Restaurant                  Park               Music Venue   
2   Middle Eastern Restaurant   Japanese Restaurant      Fast Food Restaurant   
3              Cosmetics Shop    Italian Restaurant                      Park   
4          Seafood Restaurant              Beer Bar                Restaurant   
5                 Salad Place       Bubble Tea Shop              Burger Joint   
7             Thai Restaurant                 Hotel             Deli / Bodega   
8          Italian Restaurant        Scenic Lookout                   Brewery   
9          Seafood Restaurant   Japanese Restaurant        Italian Restaurant   
10         Italian Restaurant   American Restaurant        Seafood Restaurant   
12                        Bar         Grocery Store              Burger Joint   
15         Italian Restaurant              Beer Bar                Restaurant   
16         Italian Restaurant          Gourmet Shop               Pizza Place   
17                        Gym   American Restaurant               Salad Place   
18                        Pub           Men

In [137]:
#Cluster 2
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14                 Trail            Playground          Cupcake Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14            Donut Shop      Doner Restaurant               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14   Distribution Center        Discount Store                  Diner

In [138]:
#Cluster 3
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
13  Downtown Toronto               2        Airport Lounge   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13      Airport Terminal           Coffee Shop              Boutique   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13                 Plane   Rental Car Location         Boat or Ferry   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13       Harbor / Marina      Sculpture Garden        Airport Service

In [139]:
#Cluster 4
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
11  Downtown Toronto               3                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11             Bookstore                   Bar    Italian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
11   Japanese Restaurant                Bakery           Yoga Studio   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
11              Beer Bar            Beer Store         Sandwich Place

In [140]:
#Cluster 5
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
6  Downtown Toronto               4         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                  Café                  Park            Baby Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6           Candy Store             Nightclub           Coffee Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6            Restaurant    Italian Restaurant     Athletics & Sports